### PR TITLE
DOC: Clarify how date_parser is called (GH9376)

### LINKED ISSUE #9376: read_csv: date_parser called once with arrays and then many times with strings (from each single row) as arguments
Pandas version info:

```
>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.6.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 69 Stepping 1, GenuineIntel
byteorder: little
LC_ALL: None
LANG: EN

pandas: 0.15.2
nose: 1.3.4
Cython: 0.21
numpy: 1.9.1
scipy: 0.15.0
statsmodels: 0.5.0
IPython: 2.3.1
sphinx: 1.2.3
patsy: 0.3.0
dateutil: 1.5
pytz: 2014.9
bottleneck: None
tables: 3.1.1
numexpr: 2.3.1
matplotlib: 1.4.0
openpyxl: 1.8.5
xlrd: 0.9.3
xlwt: 0.7.5
xlsxwriter: 0.5.7
lxml: 3.4.0
bs4: 4.3.2
html5lib: None
httplib2: None
apiclient: None
rpy2: None
sqlalchemy: 0.9.7
pymysql: None
psycopg2: None
```

Consider the following script, containing data to be loaded, a date parsing function, and `pd.read_csv()`:

```python
from __future__ import print_function
import StringIO
import pandas as pd
import datetime as dt

text = '''
YYYY DOY HR MN  1  2   3      4      5       6     7
2013   1  0  0 71   1 100   5571      0 99999.9 999.9
2013   1  0  1 99 999 999 999999 999999 99999.9 999.9
2013   1  0  2 71   5 100   5654     19  -348.2   9.1
2013   1  0  3 71   1 100   5647      0  -350.6   9.5
2013   1  0  4 71   1 100   5693      0  -351.2   9.4
'''

def parse_date(year, doy, hour, minute):
    print(year, type(year))
    try:  # array arguments
        year = year.astype('datetime64[Y]')
        doy = doy.astype('datetime64[D]')
        hour = hour.astype('datetime64[h]')
        minute = minute.astype('datetime64[m]')
        return year + doy + hour + minute  # return None also gives the same final DataFrame
    except:  # string arguments
        year = int(year)
        doy = int(doy)
        hour = int(hour)
        minute = int(minute)
        return dt.datetime(year, 1, 1, hour, minute) + dt.timedelta(doy - 1)


data = pd.read_csv(StringIO.StringIO(text), delim_whitespace=True,
                   date_parser=parse_date, parse_dates=[[0, 1, 2, 3]], index_col=0)
```

The `print` statement in `parse_date` outputs

```
['2013' '2013' '2013' '2013' '2013'] <type 'numpy.ndarray'>
2013 <type 'str'>
2013 <type 'str'>
2013 <type 'str'>
2013 <type 'str'>
2013 <type 'str'>
```

With `parse_date` as written above, the DataFrame is loaded properly, even when the first `return` statement in `parse_date` is changed to `return None`:

```
>>> data
                      1    2    3       4       5        6      7
YYYY_DOY_HR_MN                                                   
2013-01-01 00:00:00  71    1  100    5571       0  99999.9  999.9
2013-01-01 00:01:00  99  999  999  999999  999999  99999.9  999.9
2013-01-01 00:02:00  71    5  100    5654      19   -348.2    9.1
2013-01-01 00:03:00  71    1  100    5647       0   -350.6    9.5
2013-01-01 00:04:00  71    1  100    5693       0   -351.2    9.4

```

It strikes me as odd that the date parser is called with both whole columns and individual rows. I could not find any documentation regarding this behaviour. It immediately raises two questions in my mind:

* Why is the date parser called with both whole columns as arguments, and then called with strings from every single row as arguments? Is this a bug? It seems pointless to require the user to design for both kinds of arguments, and apparently not even use the return value from the array call.
* Wouldn't it be much better performance wise to only call the date parser once with the arrays (whole columns) as arguments, enabling the user to process the whole index in one go if possible?

### PR DESCRIPTION
**Pull request by cmeeren**
closes #9376

### CODE REVIEW

**Comment by jreback** [comment]
I can buy this. @jorisvandenbossche

**Comment by jorisvandenbossche** [comment]
Yep, looking good. Although, in reality it is still a little bit more complex, as it are actually three steps that are tried, the first and last the onces you mentioned (vectorized with the columns as input, and scalar with rows), but also vectorized on the concatenated columns into one column. Only, also adding that will make it maybe too complex?

@cmeeren Do you also want to add a similar note to the tutorial docs? Somewhere here: http://pandas.pydata.org/pandas-docs/stable/io.html#date-parsing-functions That section could also use a real example I think (of a custom defined function, not of one imported from the io.date_converters module)

**Comment by cmeeren** [comment]
@jorisvandenbossche Just to be clear concerning the second try, where you say it concatenates the columns into one column: If you use two columns, one with values `[2013, 2013, 2013]` and one with values `[1, 2, 3]`, will the second try pass the single argument `[2013, 2013, 2013, 1, 2, 3]`?

**Comment by cmeeren** [comment]
I have now mentioned the second way of calling date_parser (assuming my guess in the previous comment was correct) and added a description to the tutorial docs. I have not touched the example, since I have little experience with the Sphinx-IPython combo.

**Comment by jreback** [comment]
@cmeeren actually, I think you should mention ``pd.to_datetime()`` first. If you want to specify a format, ``read_csv`` does not currently have this implemented (it *can* infer the format however if ``infer_datetime_format=True``, but is ``False`` by default).

So it is MUCH more performant to use ``pd.to_datetime()`` AFTER parsing if you have a single column, but it needs a format specification. IOW, you should NOT use ``date_parser`` if this is the case. (In reality ``read_csv`` should just do this, but it is an open issue).

**Comment by cmeeren** [comment]
@jreback are you suggesting that I mention `pd.to_datetime()` as a possible function to use for `date_parser`?

**Comment by jreback** [comment]
no! this is in lieu of using date_parser entirely

**Comment by cmeeren** [comment]
I don't really understand the role of `pd.to_datetime()` in this scenario. Why do you suggest that I mention it in the documentation for `date_parser`? Could you give an example of how `pd.to_datetime()` might be used instead of `date_parser`?

**Comment by jorisvandenbossche** [comment]
The concatenation is like this in your example: `["2013 1", "2013 2", "2013 3"]`.

@jreback In many cases you are right about `to_datetime`, and I also think `read_csv` should take a `date_format` argument (#2586)

**Comment by jorisvandenbossche** [comment]
@jreback finishing my comment: indeed, in many case `to_datetime` will be better, but I think you specifically want to use `date_parser` when you have multiple columns that have to be combined.

BTW, you can use `pd.to_datetime` as a function for `date_parser`, no? (if you still want to do it with a one-liner)

**Comment by jreback** [comment]
so the simple heuristic is this:

- if you have multiple columns that need parsing, use ``parse_dates=[[....]]``.
- try to infer the format ``read_csv(..., infer_datetime_format=True)``
- if you have a format, the use ``date_parser=lambda x: pd.to_datetime(x, format=.....)``
- if you have a really non-standard format, finally use ``date_parser=.....``

so a naked ``date_parser`` is ALWAYS the last resort.

**Comment by jreback** [comment]
@cmeeren so what I think we should do is update the doc-string a bit (what you have is prob good). then add a section to the docs in the date parsing section giving the relative list as above.

**Comment by jorisvandenbossche** [comment]
@jreback yes, that is a nice overview of steps to follow!

The full docs on the date parsing can use an overhaul. It is now scattered a bit:

- http://pandas.pydata.org/pandas-docs/stable/io.html#specifying-date-columns
- http://pandas.pydata.org/pandas-docs/stable/io.html#date-parsing-functions
- http://pandas.pydata.org/pandas-docs/stable/io.html#inferring-datetime-format
- http://pandas.pydata.org/pandas-docs/stable/io.html#international-date-formats

where the first is not adjacent to the other three. I would have just one section with some subsections.

@cmeeren If you want, you can certainly try to tackle this! Otherwise, we just merge this as is (as it is correct information) and open a new issue for it.

**Comment by cmeeren** [comment]
First I'll correct things based on the recent feedback here. The information I added as it currently stands is not correct (specifically regarding the second "concatenation" call).

**Comment by cmeeren** [comment]
I added the list provided by @jreback. Please check the PR diff now and see if the information is good.

File: doc/source/io.rst [hunk 1]
```diff
@@ -573,6 +573,31 @@ take full advantage of the flexibility of the date parsing API:
                     date_parser=conv.parse_date_time)
    df
 
+Pandas will try to call the ``date_parser`` function in three different ways. If 
+an exception is raised, the next one is tried:
+
+1. ``date_parser`` is first called with one or more arrays as arguments,
+   as defined using `parse_dates` (e.g., ``date_parser(['2013', '2013']``, ``['1', '2'])``)
```
Comment (jorisvandenbossche) [inline doc/source/io.rst]: Here the ``` `` ``` in the middle of the date_parser(..) example should be removed I think

**Comment by jorisvandenbossche** [comment]
This looks good for me!

Apart from the one small comment, can you also squash your commits into one?

**Comment by cmeeren** [comment]
I addressed the comment and I think I managed to squash the commits now.

**Comment by jorisvandenbossche** [comment]
Thanks a lot!

**Pull request merged**